### DISCUSSION TITLE
fprime-gds customization

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1966] [Q&A] [answered]
I've implemented a custom framing protocol and I want to continue using fprime-gds to communicate with my F' instance.

Does `fprime-gds` provides a way to change the framing/deframing part through a plugin system or the only option is to create a fork of fprime-gds and modify it to meet my needs?

**Comment by SterlingPeet** [comment, marked as answer]
Unfortunately the answer is “yes and no”.  Currently the basic framework to load in additional framers exists, which you can see if you look at the pyserial connection mechanism.  However, the code required to auto-detect an external plugin does not yet exist in ``fprime-gds``. 

One of my students is working on getting this done to support CCSDS frames as they are implemented for GT-2, so it is of general interest to the community to get this going.  I won’t make promises on the delivery of this feature, but I do expect it to work before GT-2 delivers for flight (so, maybe 9 months).

**Reply by SterlingPeet** [reply]
To further clarify, the mechanism to load optional code that ``fprime-gds`` already knows *might* exist is implemented.  The mechanism to load code that ``fprime-gds`` does *not* know exists, is not currently implemented.

**Reply by LeStarch** [reply]
Is this something that will be contributed back to `fprime-gds`?  We were also doing some exploration into this, using `pluggy`.

**Reply by SterlingPeet** [reply]
We would be happy to contribute it back, under the assumption we get there first.  In this case, the student's [reasonable] proposal is to write the code that frames/deframes our CCSDS, then hook that into GDS and get it to work, and then after it works refactor to make it at ``pluggy`` plugin.  Seeing as how we are 2 weeks away from spring finals, I wouldn't count on the ``pluggy`` work occuring before August.

If the JPL team gets ``pluggy`` set up before we get there, we would just leverage it.  IMHO, ``pluggy`` is the correct framework to make the GDS extensible.

**Reply by SMorettini** [reply]
Thank you for the answer. 
We are also currently implementing CCSDS frames in Fprime, that's actually why I asked this question.

**Reply by thomas-bc** [reply]
Coming back to this discussion for posterity: we now have a plugin system in the GDS thanks to LeStarch!

https://github.com/nasa/fprime/blob/devel/docs/HowTo/develop-gds-plugins.md

So swapping `FramerDeframer` implementation should be rather straightforward

**Reply by LeStarch** [reply]
Also note, we are leaning strongly into switching to a CCSDS implementation ourselves.  Perhaps there is collaboration/coordination we should do?
